In [1]:
import pandas as pd
import os

if os.path.exists("Day11_Messy_Company_Employee_Dataset.csv"):
    df = pd.read_csv("Day11_Messy_Company_Employee_Dataset.csv")
    print("--- HEAD ---")
    print(df.head())
    print("\n--- INFO ---")
    print(df.info())
    print("\n--- MISSING ---")
    print(df.isnull().sum())
    print("\n--- DUPLICATES ---")
    print(df.duplicated().sum())
    print("\n--- DESCRIBE ---")
    print(df.describe(include='all'))
else:
    print("File not found.")

--- HEAD ---
  Employee_ID Employee_Name       Department          Job_Title   Age  Gender  \
0     EMP0098   Ananya Nair              NaN     Data Scientist  48.0   Other   
1     EMP0046   Faizan Khan        Marketing  Marketing Manager  31.0  Female   
2     EMP0017   Saira Malik  Human Resources         HR Manager  28.0  Female   
3     EMP0105    Neha Menon      Engineering  Software Engineer  30.0  Female   
4     EMP0143    Hiba Menon            Sales      Sales Manager  25.0  Female   

   Annual_Salary  Experience_Years Joining_Date       City  Performance_Score  \
0       108371.0              13.2   2017-06-13       Pune                5.0   
1       108824.0              12.1   2021-09-15     Mumbai                2.0   
2       119400.0              13.2   2021-02-05  Bengaluru                3.0   
3       111847.0               NaN   2018-08-05     Jaipur                4.0   
4        78677.0              10.9   2018-07-24  Hyderabad                4.0   

  Work_Mode  

In [2]:
print("Gender:", df['Gender'].unique())
print("Work_Mode:", df['Work_Mode'].unique())
print("Department:", df['Department'].unique())

Gender: ['Other' 'Female' 'Male' nan 'female' 'MALE']
Work_Mode: ['Office' 'Remote' 'Hybrid' nan 'REMOTE' 'remote']
Department: [nan 'Marketing' 'Human Resources' 'Engineering' 'Sales' 'Finance'
 'Data & Analytics' 'Customer Success' 'Operations' 'ENGINEERING'
 'engineering']


In [3]:
import pandas as pd
df = pd.read_csv("Day11_Messy_Company_Employee_Dataset.csv")

# 1. Drop duplicates
df.drop_duplicates(inplace=True)

# 2. Fix Categorical Inconsistencies (do this before filling NaNs so mode is accurate)
df['Gender'] = df['Gender'].str.title()
df['Work_Mode'] = df['Work_Mode'].str.title()
df['Department'] = df['Department'].str.title()

# 3. Handle Missing Values
# Numerical
df['Age'] = df['Age'].fillna(df['Age'].median())
df['Annual_Salary'] = df['Annual_Salary'].fillna(df['Annual_Salary'].median())
df['Experience_Years'] = df['Experience_Years'].fillna(df['Experience_Years'].median())
df['Performance_Score'] = df['Performance_Score'].fillna(df['Performance_Score'].mode()[0])

# Categorical
df['Department'] = df['Department'].fillna("Unknown")
df['Gender'] = df['Gender'].fillna(df['Gender'].mode()[0])
df['City'] = df['City'].fillna(df['City'].mode()[0])
df['Work_Mode'] = df['Work_Mode'].fillna(df['Work_Mode'].mode()[0])

# 4. Correct Data Types
df['Age'] = df['Age'].astype(int)
df['Joining_Date'] = pd.to_datetime(df['Joining_Date'])

print(df.isnull().sum())
print(df.info())

# 5. Export
df.to_csv("Day11_Cleaned_Company_Employee_Dataset.csv", index=False)

Employee_ID          0
Employee_Name        0
Department           0
Job_Title            0
Age                  0
Gender               0
Annual_Salary        0
Experience_Years     0
Joining_Date         0
City                 0
Performance_Score    0
Work_Mode            0
dtype: int64
<class 'pandas.core.frame.DataFrame'>
Index: 150 entries, 0 to 155
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Employee_ID        150 non-null    object        
 1   Employee_Name      150 non-null    object        
 2   Department         150 non-null    object        
 3   Job_Title          150 non-null    object        
 4   Age                150 non-null    int64         
 5   Gender             150 non-null    object        
 6   Annual_Salary      150 non-null    float64       
 7   Experience_Years   150 non-null    float64       
 8   Joining_Date       150 non-null    datetime64[ns]
 9   C

In [4]:
import pandas as pd

# 1. Load the Dataset
file_path = "Day11_Messy_Company_Employee_Dataset.csv"
df = pd.read_csv(file_path)

print("--- Dataset Condition BEFORE Cleaning ---")
print(f"Shape: {df.shape}")
print(f"Total Duplicates: {df.duplicated().sum()}")
print("Missing Values:\n", df.isnull().sum())
print("\n")

# 2. Remove Duplicate Records
# We drop identical rows to ensure each employee record is unique
df = df.drop_duplicates(keep='first')

# 3. Standardize Inconsistent Categorical Entries
# Converting text to Title Case resolves inconsistencies like "MALE", "female", "REMOTE", "remote", etc.
df['Gender'] = df['Gender'].str.title()
df['Work_Mode'] = df['Work_Mode'].str.title()
df['Department'] = df['Department'].str.title()

# 4. Handle Missing Values (Imputation)
# Numerical Data: We use median to avoid outliers skewing the replacement values
df['Age'] = df['Age'].fillna(df['Age'].median())
df['Annual_Salary'] = df['Annual_Salary'].fillna(df['Annual_Salary'].median())
df['Experience_Years'] = df['Experience_Years'].fillna(df['Experience_Years'].median())

# For Performance_Score (which acts like a categorical rating 1-5), mode is best
df['Performance_Score'] = df['Performance_Score'].fillna(df['Performance_Score'].mode()[0])

# Categorical Data: We use mode (most frequent value) or a placeholder
df['Gender'] = df['Gender'].fillna(df['Gender'].mode()[0])
df['City'] = df['City'].fillna(df['City'].mode()[0])
df['Work_Mode'] = df['Work_Mode'].fillna(df['Work_Mode'].mode()[0])
df['Department'] = df['Department'].fillna("Unknown") # Better to flag unknown departments

# 5. Fix Data Types
# Age should be a whole number (integer)
df['Age'] = df['Age'].astype(int)

# Joining_Date must be a datetime object instead of a string
df['Joining_Date'] = pd.to_datetime(df['Joining_Date'])

print("--- Dataset Condition AFTER Cleaning ---")
print(f"Shape: {df.shape}")
print(f"Total Duplicates: {df.duplicated().sum()}")
print("Missing Values:\n", df.isnull().sum())
print("\n")

# 6. Export the Cleaned Dataset
output_filename = "Day11_Cleaned_Company_Employee_Dataset.csv"
df.to_csv(output_filename, index=False)
print(f"Success! Cleaned data exported to: {output_filename}")

--- Dataset Condition BEFORE Cleaning ---
Shape: (157, 12)
Total Duplicates: 7
Missing Values:
 Employee_ID          0
Employee_Name        0
Department           5
Job_Title            0
Age                  4
Gender               5
Annual_Salary        5
Experience_Years     3
Joining_Date         0
City                 5
Performance_Score    3
Work_Mode            2
dtype: int64


--- Dataset Condition AFTER Cleaning ---
Shape: (150, 12)
Total Duplicates: 0
Missing Values:
 Employee_ID          0
Employee_Name        0
Department           0
Job_Title            0
Age                  0
Gender               0
Annual_Salary        0
Experience_Years     0
Joining_Date         0
City                 0
Performance_Score    0
Work_Mode            0
dtype: int64


Success! Cleaned data exported to: Day11_Cleaned_Company_Employee_Dataset.csv


### Data Cleaning Summary

Here are the observations and decisions made during the data-cleaning process:

*   **Duplicate Records:** The initial dataset contained 157 rows, but exactly **7 duplicate rows** were identified. The `drop_duplicates()` function was used to remove them, bringing the count down to 150 unique records.
*   **Inconsistent Text Formats:** Columns like `Gender`, `Work_Mode`, and `Department` suffered from inconsistent capitalization (e.g., mixing "MALE", "female", and "Male"). The `.str.title()` method was applied to convert all entries to Title Case, harmonizing the categories before evaluating their missing values.
*   **Numerical Missing Values:** Missing entries in `Age`, `Annual_Salary`, and `Experience_Years` were imputed using `.fillna()` combined with the **median**. The median was chosen over the mean because it is robust against unusually high or low outlier values (which are common in salary and age data).
*   **Categorical Missing Values:** Missing entries in `Gender`, `City`, `Work_Mode`, and `Performance_Score` were filled using the **mode** (the most frequently occurring value in each respective column). For the `Department` column, missing values were filled with the string `"Unknown"` to avoid assigning an employee to a random department without evidence.
*   **Correcting Data Types:**
    *   The `Age` column was converted from floating-point decimals to an **integer** (`int`) using `.astype()`, as age is represented in whole years.
    *   The `Joining_Date` column was originally read as an `object` (string). It was successfully converted to a proper **datetime** structure using `pd.to_datetime()`, allowing for future time-based filtering.